tiatoolbox/models/architecture/sccnn.py

```diff
@@ -0,0 +1,372 @@
+        module_dict["dropout1"] = nn.Dropout2d(p=0.5)
+        module_dict["l5"] = conv_act_block(1024, 512, 1)
+        module_dict["dropout2"] = nn.Dropout2d(p=0.5)
+        module_dict["sc"] = spatially_constrained_layer1(512, 3)
+
+        self.layer = nn.ModuleDict(module_dict)
+
+    def spatially_constrained_layer2(self, sc1_0, sc1_1, sc1_2) -> torch.Tensor:
+        """Spatially constrained layer 2.
+
+        Estimates row, column and height for sc2 layer mapping.
+
+        Args:
+            sc1_0 (torch.Tensor):
+                Output of spatially_constrained_layer1 estimating
+                the x position of the nucleus.
+            sc1_1 (int):
+                Output of spatially_constrained_layer1 estimating
+                the y position of the nucleus.
+            sc1_2 (int):
+                Output of spatially_constrained_layer1 estimating
+                the confidence in nucleus detection.
+
+        Returns:
+            :class:`torch.Tensor`:
+                Probability map using the estimates from
+                spatially_constrained_layer1.
+
+        """
+        x = torch.tile(self.xv, dims=[sc1_0.size(0), 1, 1, 1])  # Tile for batch size
+        y = torch.tile(self.yv, dims=[sc1_0.size(0), 1, 1, 1])
+        xvr = (x - sc1_0) ** 2
+        yvc = (y - sc1_1) ** 2
+        out_map = xvr + yvc
+        out_map_threshold = torch.lt(out_map, self.radius).type(torch.float32)
+        denominator = 1 + (out_map / 2)
+        sc2 = sc1_2 / denominator
+        return sc2 * out_map_threshold
+
+    def forward(self, input_tensor: torch.Tensor):  # skipcq: PYL-W0221
+        """Logic for using layers defined in init.
+
+        This method defines how layers are used in forward operation.
+
+        Args:
+            input_tensor (torch.Tensor):
+                Input images, the tensor is in the shape of NCHW.
+
+        Returns:
+            torch.Tensor:
+                Output map for cell detection. Peak detection should be applied
+                to this output for cell detection.
+
+        """
+
+        def spatially_constrained_layer1(
+            layer: torch.nn.Module,
+            in_tensor: torch.Tensor,
+            out_height: int = 13,
+            out_width: int = 13,
+        ) -> tuple:
+            """Spatially constrained layer 1.
+
+            Estimates row, column and height for
+            spatially_constrained_layer2 layer mapping.
+
+            Args:
+                layer (torch.nn.Module):
+                    Torch layer as ModuleDict.
+                in_tensor (torch.Tensor):
+                    Input Tensor.
+                out_height (int):
+                    Output height.
+                out_width (int):
+                    Output Width
+
+            Returns:
+                tuple of :class:`torch.Tensor`:
+                    Row, Column and height estimates used for
+                    spatially_constrained_layer2 mapping.
```

Comment:
Might be nice to use the bulleted list format which we use elsewhere in the toolbox for tuple returns.

Reply:
Not required here as these are single values.